File: assets/js/permissions.js
/* ==========================================================================
   permissions.js — Phase 4 of the AuditPulse Access Portal integration
   (see AuditPulse-Access/README.md; Phase 3 is api/access_management.py
   on the backend, which is what puts `role` and `permissions` on the
   logged-in user in the first place).

   Reads the current user's role (assets/js/api.js's Api.auth.getUser(),
   which now carries `role` and a computed `permissions` module map
   straight from the backend's UserOut — see schemas/user.py) and:
     - hides sidebar links / call-to-action buttons for modules the role
       doesn't have,
     - redirects away from a gated page reached by typing its URL
       directly, with an explanatory toast.

   This is the UX layer only. The actual enforcement is server-side
   (backend/config/permissions.py's require_module, applied to the
   scheduler/settings routers and the audit-create route) — hiding a
   button here never stops a direct API call, on purpose: a person
   editing localStorage or calling the API by hand shouldn't be able to
   grant themselves a module this file can't actually take away.
   ========================================================================== */

window.Permissions = (function () {
  var U = window.Utils;

  // Mirrors backend/config/permissions.py's MODULE_PERMISSIONS. Only used
  // as a fallback — for a session stored before this shipped, whose
  // cached user has a `role` but no `permissions` field yet. Once that
  // session refreshes (next login) the backend's own copy takes over.
  // Keep both in sync if a role's access ever changes.
  var FALLBACK_MODULE_PERMISSIONS = {
    Admin:    { dashboard: true, audits: true,  analytics: true, reports: true, scheduler: true,  settings: true },
    Auditor:  { dashboard: true, audits: true,  analytics: true, reports: true, scheduler: true,  settings: false },
    Reviewer: { dashboard: true, audits: true,  analytics: true, reports: true, scheduler: false, settings: false },
    Viewer:   { dashboard: true, audits: false, analytics: true, reports: true, scheduler: false, settings: false }
  };

  // Which page (by filename) belongs to which gated module. A page with
  // no entry here (dashboard.html, report.html, history.html) is open to
  // every role — matches the backend matrix, which has no "history" key
  // and marks dashboard/reports/analytics true for all four roles.
  var PAGE_MODULES = {
    'audit.html': 'audits',
    'scheduler.html': 'scheduler',
    'settings.html': 'settings'
  };

  var MODULE_LABELS = {
    audits: 'New Audit',
    scheduler: 'Scheduler',
    settings: 'Settings'
  };

  var DENIED_KEY = 'ap_access_denied_module';

  function currentPage() {
    return (location.pathname.split('/').pop() || 'dashboard.html');
  }

  function permissionsFor(user) {
    if (!user) return null;
    if (user.permissions) return user.permissions;
    return FALLBACK_MODULE_PERMISSIONS[user.role] || FALLBACK_MODULE_PERMISSIONS.Viewer;
  }

  // Whether the logged-in user's role has `module`. Modules the matrix
  // doesn't mention (e.g. "history") are treated as open to everyone,
  // and so is anything before login resolves — a missing session isn't
  // this module's problem to enforce, the page's own auth handling is.
  function can(module) {
    var user = window.Api && window.Api.auth.getUser();
    var perms = permissionsFor(user);
    if (!perms) return true;
    if (!(module in perms)) return true;
    return !!perms[module];
  }

  function hrefModule(href) {
    if (!href) return null;
    var base = href.split('#')[0].split('?')[0];
    return PAGE_MODULES[base] || null;
  }

  // Hides (not just disables) any link pointing at a gated page the
  // current role can't use — sidebar nav items plus any other
  // call-to-action to the same page (e.g. dashboard.html's "New Audit"
  // hero button and quick-action card). Walks every <a> on the page
  // rather than requiring a data-attribute on each one, so a future CTA
  // pointing at scheduler.html/settings.html/audit.html is covered for
  // free without needing to remember to tag it.
  function hideGatedLinks() {
    U.qsa('a[href]').forEach(function (link) {
      var module = hrefModule(link.getAttribute('href'));
      if (!module || can(module)) return;
      var item = link.closest('li') || link;
      item.style.display = 'none';
    });
  }

  // Direct-URL safety net: hiding a nav link doesn't stop someone typing
  // scheduler.html into the address bar. Bounces them to the dashboard
  // with an explanatory toast (shown there — see showDeniedNoticeIfAny)
  // rather than letting them sit on a page whose data calls are about to
  // 403 anyway. Returns true if it redirected, so init() below knows not
  // to bother running anything else on a page that's navigating away.
  function redirectIfDenied() {
    var module = PAGE_MODULES[currentPage()];
    if (!module || can(module)) return false;

    try { window.sessionStorage.setItem(DENIED_KEY, module); } catch (e) { /* ignore */ }
    window.location.replace('dashboard.html');
    return true;
  }

  // Shows the one-shot toast left behind by redirectIfDenied() once
  // landed back on an ungated page (in practice always dashboard.html,
  // its redirect target).
  function showDeniedNoticeIfAny() {
    var module;
    try { module = window.sessionStorage.getItem(DENIED_KEY); } catch (e) { module = null; }
    if (!module) return;
    try { window.sessionStorage.removeItem(DENIED_KEY); } catch (e) { /* ignore */ }
    if (!window.Notifications) return;

    var user = window.Api && window.Api.auth.getUser();
    var role = (user && user.role) || 'your role';
    window.Notifications.warning(
      'Access restricted',
      (MODULE_LABELS[module] || module) + ' isn\u2019t available for the ' + role +
        ' role. An org admin can change this in the Access Portal.'
    );
  }

  // Small markup snippet for a "Role: X" line — app.js's initProfileMenu
  // builds the profile dropdown itself and appends this to it. This is
  // the one place Phase 4 makes the role AuditPulse read back visible to
  // the person it belongs to, rather than something only silently
  // driving what they can click.
  function roleBadgeHtml(role) {
    return '<div style="padding:2px 10px 8px; color:var(--text-tertiary); font-size:11px; ' +
      'text-transform:uppercase; letter-spacing:.04em;">Role: ' + U.escapeHtml(role || 'Viewer') + '</div>';
  }

  function init() {
    var user = window.Api && window.Api.auth.getUser();
    if (!user) return; // no session yet — nothing to gate until one exists

    if (redirectIfDenied()) return;
    hideGatedLinks();
    showDeniedNoticeIfAny();
  }

  return {
    can: can,
    init: init,
    roleBadgeHtml: roleBadgeHtml,
    PAGE_MODULES: PAGE_MODULES
  };
})();

File: backend/config/permissions.py
"""
config/permissions.py

Phase 4 of the AuditPulse Access Portal integration (see
AuditPulse-Access/README.md and api/access_management.py for Phase 3).

Phase 3 made AuditPulse *store* the role the Access Portal assigns
(`User.role`). This module is what makes AuditPulse actually *read* it:
a role -> module permission matrix, plus a FastAPI dependency that gates
a route on it.

MODULE_PERMISSIONS is a deliberate copy of AuditPulse-Access/backend/
roles.py's `PERMISSIONS` dict, not an import of it — the two are separate
deployments (this backend has no dependency on the portal's repo/runtime),
so the portal stays the single *authored* source of truth for what each
role can do, and this is AuditPulse's own copy of that same decision.
If a role is ever added or a module's access changes, update both.

Kept deliberately dumb, matching roles.py's own comment: a plain dict,
no `permissions` table, no per-organisation overrides. Add those to both
sides together if a customer ever needs a role's access customised.
"""

from typing import Dict

from fastapi import Depends, HTTPException, status

# Module -> allowed, per role name. Same four fixed roles and six modules
# as AuditPulse-Access/backend/roles.py's PERMISSIONS. "history" isn't a
# gated module here either, on purpose (see that file's comment) — every
# role can see their own activity log.
MODULE_PERMISSIONS: Dict[str, Dict[str, bool]] = {
    "Admin":    {"dashboard": True, "audits": True,  "analytics": True, "reports": True, "scheduler": True,  "settings": True},
    "Auditor":  {"dashboard": True, "audits": True,  "analytics": True, "reports": True, "scheduler": True,  "settings": False},
    "Reviewer": {"dashboard": True, "audits": True,  "analytics": True, "reports": True, "scheduler": False, "settings": False},
    "Viewer":   {"dashboard": True, "audits": False, "analytics": True, "reports": True, "scheduler": False, "settings": False},
}

# Any role not in the dict above (shouldn't happen once Phase 3's
# migration backfills "Viewer", but a hand-edited row or a future role
# the portal defines that this copy hasn't caught up with yet is
# possible) gets the least-privileged existing role's permissions rather
# than a KeyError or, worse, defaulting to allow-everything.
FALLBACK_ROLE = "Viewer"


def get_permissions(role: str) -> Dict[str, bool]:
    """Module access dict for a role name. Never raises — an unknown role
    just falls back to Viewer's (most restrictive) permissions."""
    return MODULE_PERMISSIONS.get(role, MODULE_PERMISSIONS[FALLBACK_ROLE])


def require_module(module: str):
    """Dependency factory: `Depends(require_module("scheduler"))` in place
    of `Depends(get_current_user)` on a route. Resolves the current user
    exactly as before (still 401s if unauthenticated) and additionally
    403s if that user's role doesn't have `module` enabled.

    This is the real enforcement — the frontend hiding a sidebar link or
    redirecting away from a page (assets/js/permissions.js) is only the
    UX layer on top of it, never a substitute for it, since a hidden
    button doesn't stop a direct API call.
    """
    # Imported inside the factory, not at module load, to avoid a circular
    # import: api/auth.py doesn't import this module, but several routers
    # that *do* import this module (api/audit.py, api/scheduler.py,
    # api/settings.py) already import get_current_user from api.auth
    # too, so this keeps the import graph a simple one-way dependency on
    # api.auth rather than a cycle through it.
    from api.auth import User, get_current_user

    async def _dependency(current_user: User = Depends(get_current_user)) -> User:
        if not get_permissions(current_user.role).get(module, False):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=(
                    f"Your role ({current_user.role}) doesn't have access "
                    f"to {module}. Ask an org admin to change your role in "
                    "the Access Portal if this is wrong."
                ),
            )
        return current_user

    return _dependency
